src/QMCHamiltonians/CoulombPBCAB.cpp

```diff
@@ -633,39 +767,57 @@ CoulombPBCAB::evalLRwithForces(ParticleSet& P)
   const StructFact& RhoKA(*(PtclA.SK));
   const StructFact& RhoKB(*(P.SK));
   std::vector<TinyVector<RealType,DIM> > grad(PtclA.getTotalNum());
-  for(int j=0; j<NumSpeciesB; j++)
+ for(int j=0; j<NumSpeciesB; j++)
   {
     for (int iat=0; iat<grad.size(); iat++)
       grad[iat] = TinyVector<RealType,DIM>(0.0, 0.0, 0.0);
-    AB->evaluateGrad(PtclA, P, j, Zat, grad);
+    dAB->evaluateGrad(PtclA, P, j, Zat, grad);
     for (int iat=0; iat<grad.size(); iat++)
-      forces[iat] += Qspec[j]*grad[iat];
+      forces[iat] += Qspec[j]*Zat[iat]*grad[iat];
   } // electron species
   return evalLR(P);
 }
 
 CoulombPBCAB::Return_t
 CoulombPBCAB::evalSRwithForces(ParticleSet& P)
 {
+  CONSTEXPR mRealType czero(0);
   const DistanceTableData &d_ab(*P.DistTables[myTableIndex]);
-  mRealType res=0.0;
-  //Loop over distinct eln-ion pairs
-  for(int iat=0; iat<NptclA; iat++)
+  mRealType res=czero;
+  //Temporary variables for computing energy and forces.
+  mRealType rV(0);
+  mRealType frV(0),fdrV(0);
+  mRealType rinv(1.0);
+  //Magnitude of force.
+  mRealType dvdr(0.0);
+  if(d_ab.DTType == DT_SOA)
   {
-    mRealType esum = 0.0;
-    RadFunctorType* rVs=Vat[iat];
-    for(int nn=d_ab.M[iat], jat=0; nn<d_ab.M[iat+1]; ++nn,++jat)
+    for(size_t b=0; b<NptclB; ++b)
     {
-      RealType rV, d_rV_dr, d2_rV_dr2, V;
-      rV = rVs->splint(d_ab.r(nn), d_rV_dr, d2_rV_dr2);
-      V = rV *d_ab.rinv(nn);
-      PosType drhat = d_ab.rinv(nn) * d_ab.dr(nn);
-      esum += Qat[jat]*d_ab.rinv(nn)*rV;
-      forces[iat] += Zat[iat]*Qat[jat] * //g(d_ab.r(nn)) *
-                     (d_rV_dr - V)*d_ab.rinv(nn) *drhat;
+      const RealType* restrict dist=d_ab.Distances[b];
+      RowContainerType dr = d_ab.Displacements[b];
+      mRealType esum=czero;
+      for(size_t a=0; a<NptclA; ++a)
+      {
+        //Low hanging SIMD fruit here.  See J1/J2 grad computation.
+        //Too lazy for now.  
```

Comment:
This is good to leave hints/hooks in the code. In this place, doing SIMD is non trivial. No problem leaving such message and we can always revisit later.